pkgs/top-level/all-packages.nix

```diff
@@ -11298,6 +11298,14 @@ let
 
   cutecom = callPackage ../tools/misc/cutecom { };
 
+  cutegram =
+    let f = ../applications/networking/instant-messengers/telegram;
+        cp = qt5Libs.callPackage;
+    in cp "${f}/cutegram" rec {
```

Comment:
> I’m not sure how this is any different from replacing every cp with callPackage and every f with ../applications/networking/instant-messengers/telegram. Isn’t nix referentially transparent? 

Nix is referentially transparent, but you have to consider what you are referring to in each case. By that I mean, paths are not strings, they have their own, distinct type! Because Nix requires all build inputs to live in the Nix store, converting a path to a string copies the path to the store and returns the string *of that store path*. The evaluation fails because the copying doesn't actually take place until the package is built. That can't happen until evaluation finishes, but evaluating `callPackage` requires the path to actually exist. This is all referentially transparent: if you use the same inputs, the same store path will be created.

> Also, I’m not using import anywhere. I don’t really understand what goes wrong here.

`callPackage` is just a combination of `import`, `makeOverridable`, and a few tricks.

> I chose this semi-top-level style because these libraries are only really used by cutegram in nixpkgs at the moment, but there might be other packages who use it, and since they are only libraries I don’t want to clutter the top-level.

To really be useful to other packages, those need to be added to `qt5LibsFun` anyway, so it's not clutter.